File: packages/localization/android/src/main/java/io/capawesome/capacitorjs/plugins/localization/Localization.java
package io.capawesome.capacitorjs.plugins.localization;

import android.content.Context;
import android.os.LocaleList;
import android.text.TextUtils;
import android.text.format.DateFormat;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.capawesome.capacitorjs.plugins.localization.classes.results.GetLocalesResult;
import io.capawesome.capacitorjs.plugins.localization.classes.results.GetSettingsResult;
import io.capawesome.capacitorjs.plugins.localization.classes.results.LocaleResult;
import io.capawesome.capacitorjs.plugins.localization.interfaces.NonEmptyResultCallback;
import java.text.DecimalFormatSymbols;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Currency;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

public class Localization {

    @NonNull
    private final Context context;

    public Localization(@NonNull Context context) {
        this.context = context;
    }

    public void getLocales(@NonNull NonEmptyResultCallback<GetLocalesResult> callback) {
        LocaleList localeList = LocaleList.getDefault();
        List<LocaleResult> locales = new ArrayList<>();
        for (int i = 0; i < localeList.size(); i++) {
            locales.add(createLocaleResult(localeList.get(i)));
        }
        callback.success(new GetLocalesResult(locales));
    }

    public void getSettings(@NonNull NonEmptyResultCallback<GetSettingsResult> callback) {
        String timeZone = TimeZone.getDefault().getID();
        boolean uses24HourClock = DateFormat.is24HourFormat(context);
        int firstDayOfWeek = convertToIsoDayOfWeek(Calendar.getInstance().getFirstDayOfWeek());
        callback.success(new GetSettingsResult(timeZone, uses24HourClock, firstDayOfWeek));
    }

    private int convertToIsoDayOfWeek(int calendarDayOfWeek) {
        return ((calendarDayOfWeek + 5) % 7) + 1;
    }

    @NonNull
    private LocaleResult createLocaleResult(@NonNull Locale locale) {
        String languageTag = locale.toLanguageTag();
        String languageCode = locale.getLanguage();
        String regionCode = locale.getCountry().isEmpty() ? null : locale.getCountry();
        Currency currency = resolveCurrency(locale);
        String currencyCode = currency == null ? null : currency.getCurrencyCode();
        String currencySymbol = currency == null ? null : currency.getSymbol(locale);
        DecimalFormatSymbols symbols = DecimalFormatSymbols.getInstance(locale);
        String decimalSeparator = String.valueOf(symbols.getDecimalSeparator());
        String groupingSeparator = String.valueOf(symbols.getGroupingSeparator());
        String textDirection = resolveTextDirection(locale);
        String measurementSystem = resolveMeasurementSystem(regionCode);
        return new LocaleResult(
            languageTag,
            languageCode,
            regionCode,
            currencyCode,
            currencySymbol,
            decimalSeparator,
            groupingSeparator,
            textDirection,
            measurementSystem
        );
    }

    @Nullable
    private Currency resolveCurrency(@NonNull Locale locale) {
        try {
            return Currency.getInstance(locale);
        } catch (IllegalArgumentException exception) {
            return null;
        }
    }

    @Nullable
    private String resolveMeasurementSystem(@Nullable String regionCode) {
        if (regionCode == null) {
            return null;
        }
        switch (regionCode.toUpperCase(Locale.ROOT)) {
            case "US":
            case "LR":
            case "MM":
                return "us";
            case "GB":
                return "uk";
            default:
                return "metric";
        }
    }

    @NonNull
    private String resolveTextDirection(@NonNull Locale locale) {
        int layoutDirection = TextUtils.getLayoutDirectionFromLocale(locale);
        return layoutDirection == View.LAYOUT_DIRECTION_RTL ? "rtl" : "ltr";
    }
}

File: packages/localization/ios/Plugin/Localization.swift
import Foundation

@objc public class Localization: NSObject {
    let plugin: LocalizationPlugin

    init(plugin: LocalizationPlugin) {
        self.plugin = plugin
    }

    @objc public func getLocales(completion: @escaping (GetLocalesResult?, Error?) -> Void) {
        let locales = Locale.preferredLanguages.map { createLocaleResult(identifier: $0) }
        completion(GetLocalesResult(locales: locales), nil)
    }

    @objc public func getSettings(completion: @escaping (GetSettingsResult?, Error?) -> Void) {
        let timeZone = TimeZone.current.identifier
        let uses24HourClock = determineUses24HourClock()
        let firstDayOfWeek = convertToIsoDayOfWeek(Calendar.current.firstWeekday)
        let result = GetSettingsResult(timeZone: timeZone, uses24HourClock: uses24HourClock, firstDayOfWeek: firstDayOfWeek)
        completion(result, nil)
    }

    private func convertToIsoDayOfWeek(_ weekday: Int) -> Int {
        return ((weekday + 5) % 7) + 1
    }

    private func createLocaleResult(identifier: String) -> LocaleResult {
        let locale = Locale(identifier: identifier)
        let languageCode: String?
        let regionCode: String?
        let currencyCode: String?
        if #available(iOS 16, *) {
            languageCode = locale.language.languageCode?.identifier
            regionCode = locale.region?.identifier
            currencyCode = locale.currency?.identifier
        } else {
            languageCode = locale.languageCode
            regionCode = locale.regionCode
            currencyCode = locale.currencyCode
        }
        let currencySymbol = locale.currencySymbol.flatMap { $0.isEmpty ? nil : $0 }
        let resolvedLanguageCode = languageCode ?? identifier
        return LocaleResult(
            languageTag: identifier,
            languageCode: resolvedLanguageCode,
            regionCode: regionCode,
            currencyCode: currencyCode,
            currencySymbol: currencySymbol,
            decimalSeparator: locale.decimalSeparator,
            groupingSeparator: locale.groupingSeparator,
            textDirection: resolveTextDirection(languageCode: resolvedLanguageCode),
            measurementSystem: resolveMeasurementSystem(locale: locale, regionCode: regionCode)
        )
    }

    private func determineUses24HourClock() -> Bool {
        guard let format = DateFormatter.dateFormat(fromTemplate: "j", options: 0, locale: Locale.current) else {
            return false
        }
        return !format.contains("a")
    }

    private func resolveMeasurementSystem(locale: Locale, regionCode: String?) -> String? {
        if #available(iOS 16, *) {
            switch locale.measurementSystem {
            case .metric:
                return "metric"
            case .us:
                return "us"
            case .uk:
                return "uk"
            default:
                return nil
            }
        }
        guard let regionCode = regionCode else {
            return nil
        }
        switch regionCode.uppercased() {
        case "US", "LR", "MM":
            return "us"
        case "GB":
            return "uk"
        default:
            return "metric"
        }
    }

    private func resolveTextDirection(languageCode: String) -> String {
        return Locale.characterDirection(forLanguage: languageCode) == .rightToLeft ? "rtl" : "ltr"
    }
}
